modules/internal/ChapelArray.chpl

```diff
@@ -2037,7 +2061,18 @@ module ChapelArray {
             hi = this.domain.high+1;
       const newDom = {lo..hi};
       on this._value {
-        this._value.dsiReallocate(newDom);
+        if !this._value.dataAllocRange.member(hi) {
+          /* The new index is not in the allocated space.  We'll need to
+             realloc it. */
+          if this._value.dataAllocRange.length < this.domain.numIndices {
```

Comment:
One way or another, it ought to be factored into a separate function, to improve readability of the code.